<a href="https://colab.research.google.com/github/Jandsy/ml_finance_imperial/blob/main/Coursework/CourseWork.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **<center>Machine Learning and Finance </center>**


## <center> CourseWork 2024 - StatArb </center>



In this coursework, you will delve into and replicate selected elements of the research detailed in the paper **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**. **However, we will not reproduce the entire study.**

## Overview

This study redefines Statistical Arbitrage (StatArb) by combining Autoencoder architectures and policy learning to generate trading strategies. Traditionally, StatArb involves finding the mean of a synthetic asset through classical or PCA-based methods before developing a mean reversion strategy. However, this paper proposes a data-driven approach using an Autoencoder trained on US stock returns, integrated into a neural network representing portfolio trading policies to output portfolio allocations directly.


## Coursework Goal

This coursework will replicate these results, providing hands-on experience in implementing and evaluating this innovative end-to-end policy learning Autoencoder within financial trading strategies.

## Outline

- [Data Preparation and Exploration](#Data-Preparation-and-Exploration)
- [Fama French Analysis](#Fama-French-Analysis)
- [PCA Analysis](#PCA-Analysis)
- [Ornstein Uhlenbeck](#Ornstein-Uhlenbeck)
- [Autoencoder Analysis](#Autoencoder-Analysis)



**Description:**
The Coursework is graded on a 100 point scale and is divided into five  parts. Below is the mark distribution for each question:

| **Problem**  | **Question**          | **Number of Marks** |
|--------------|-----------------------|---------------------|
| **Part A**   | Question 1            | 4                   |
|              | Question 2            | 1                   |
|              | Question 3            | 3                   |
|              | Question 4            | 3                   |
|              | Question 5            | 1                   |
|              | Question 6            | 3                   |
|**Part  B**    | Question 7           | 1                   |
|              | Question 8            | 5                   |
|              | Question 9            | 4                   |
|              | Question 10           | 5                   |
|              | Question 11           | 2                   |
|              | Question 12           | 3                   |
|**Part  C**    | Question 13          | 3                   |
|              | Question 14           | 1                   |
|              | Question 15           | 3                   |
|              | Question 16           | 2                   |
|              | Question 17           | 7                   |
|              | Question 18           | 6                   |
|              | Question 19           | 3                   |
|  **Part  D** | Question 20           | 3                   |
|              | Question 21           | 5                   |
|              | Question 22           | 2                   |
|  **Part  E** | Question 23           | 2                   |
|              | Question 24           | 1                   |
|              | Question 25           | 3                   |
|              | Question 26           | 10                  |
|              | Question 27           | 1                   |
|              | Question 28           | 3                   |
|              | Question 29           | 3                   |
|              | Question 30           | 7                   |




Please read the questions carefully and do your best. Good luck!

## Objectives



## 1. Data Preparation and Exploration
Collect, clean, and prepare US stock return data for analysis.

## 2. Fama French Analysis
Utilize Fama French Factors to isolate the idiosyncratic components of stock returns, differentiating them from market-wide effects. This analysis helps in understanding the unique characteristics of individual stocks relative to broader market trends.

## 3. PCA Analysis
Employ Principal Component Analysis (PCA) to identify hidden structures and reduce dimensionality in the data. This method helps in extracting significant patterns that might be obscured in high-dimensional datasets.

## 4. Ornstein-Uhlenbeck Process
Analyze mean-reverting behavior in stock prices using the Ornstein-Uhlenbeck process. This stochastic process is useful for modeling and forecasting based on the assumption that prices will revert to a long-term mean.

## 5. Building a Basic Autoencoder Model
Construct and train a standard Autoencoder to extract residual idiosyncratic risk.








# Data Preparation and Exploration


---
<font color=green>Q1: (4 Marks)</font>
<br><font color='green'>
Write a Python function that accepts a URL parameter and retrieves the NASDAQ-100 companies and their ticker symbols by scraping the relevant Wikipedia page using **[Requests](https://pypi.org/project/requests/)** and **[BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)**. Your function should return the data as a list of tuples, with each tuple containing the company name and its ticker symbol. Then, call your function with the appropriate Wikipedia page URL and print the data in a 'Company: Ticker' format.

</font>

---


In [1]:
import requests
from bs4 import BeautifulSoup

# Function to get NASDAQ-100 companies and their tickers
def get_nasdaq100_companies(nasdaq100):
    # Send a GET request to the provided URL
    response = requests.get(nasdaq100)
    
    # Check if the request was successful
    if response.status_code == 200:
        # Parse the HTML content of the response using BeautifulSoup
        soup = BeautifulSoup(response.content, 'html.parser')
        # Find the table with the ID 'constituents'
        table = soup.find('table', {'id': 'constituents'})
        
        # Extract and return the company names and tickers as a tuple
        return tuple((row.find_all('td')[0].text.strip(), row.find_all('td')[1].text.strip())
                     for row in table.find_all('tr')[1:])
    else:
        # Print an error message if the request failed
        print("Failed to retrieve data from the URL.")
        return None

# Wikipedia URL containing NASDAQ-100 companies
wiki_url = "https://en.wikipedia.org/wiki/NASDAQ-100"
# Get the NASDAQ-100 companies and their tickers
nasdaq100_data = get_nasdaq100_companies(wiki_url)

# If data was successfully retrieved, print each company and its ticker
if nasdaq100_data:
    # Create a list of tickers for further use if needed
    tickers_list = [company[1] for company in nasdaq100_data]
    for company in nasdaq100_data:
        print(f"{company[0]}: {company[1]}")

# Double-check if the dataset is in tuple format
if isinstance(nasdaq100_data, tuple):
    print("The dataset is a tuple.")
else:
    print("The dataset is not a tuple.")

Adobe Inc.: ADBE
ADP: ADP
Airbnb: ABNB
Alphabet Inc. (Class A): GOOGL
Alphabet Inc. (Class C): GOOG
Amazon: AMZN
Advanced Micro Devices Inc.: AMD
American Electric Power: AEP
Amgen: AMGN
Analog Devices: ADI
Ansys: ANSS
Apple Inc.: AAPL
Applied Materials: AMAT
ASML Holding: ASML
AstraZeneca: AZN
Atlassian: TEAM
Autodesk: ADSK
Baker Hughes: BKR
Biogen: BIIB
Booking Holdings: BKNG
Broadcom Inc.: AVGO
Cadence Design Systems: CDNS
CDW Corporation: CDW
Charter Communications: CHTR
Cintas: CTAS
Cisco: CSCO
Coca-Cola Europacific Partners: CCEP
Cognizant: CTSH
Comcast: CMCSA
Constellation Energy: CEG
Copart: CPRT
CoStar Group: CSGP
Costco: COST
CrowdStrike: CRWD
CSX Corporation: CSX
Datadog: DDOG
DexCom: DXCM
Diamondback Energy: FANG
Dollar Tree: DLTR
DoorDash: DASH
Electronic Arts: EA
Exelon: EXC
Fastenal: FAST
Fortinet: FTNT
GE HealthCare: GEHC
Gilead Sciences: GILD
GlobalFoundries: GFS
Honeywell: HON
Idexx Laboratories: IDXX
Illumina, Inc.: ILMN
Intel: INTC
Intuit: INTU
Intuitive Surgical: I

---
<font color=green>Q2: (1 Mark)</font>
<br><font color='green'>
Given a list of tuples representing NASDAQ-100 companies (where each tuple contains a company name and its ticker symbol), write a Python script to extract all ticker symbols into a separate list called `tickers_list`.
</font>
---


In [2]:
#Q2

tickers_list = [company[1] for company in nasdaq100_data]

print(tickers_list)

#Double check if all companies are called
print(len(tickers_list))

['ADBE', 'ADP', 'ABNB', 'GOOGL', 'GOOG', 'AMZN', 'AMD', 'AEP', 'AMGN', 'ADI', 'ANSS', 'AAPL', 'AMAT', 'ASML', 'AZN', 'TEAM', 'ADSK', 'BKR', 'BIIB', 'BKNG', 'AVGO', 'CDNS', 'CDW', 'CHTR', 'CTAS', 'CSCO', 'CCEP', 'CTSH', 'CMCSA', 'CEG', 'CPRT', 'CSGP', 'COST', 'CRWD', 'CSX', 'DDOG', 'DXCM', 'FANG', 'DLTR', 'DASH', 'EA', 'EXC', 'FAST', 'FTNT', 'GEHC', 'GILD', 'GFS', 'HON', 'IDXX', 'ILMN', 'INTC', 'INTU', 'ISRG', 'KDP', 'KLAC', 'KHC', 'LRCX', 'LIN', 'LULU', 'MAR', 'MRVL', 'MELI', 'META', 'MCHP', 'MU', 'MSFT', 'MRNA', 'MDLZ', 'MDB', 'MNST', 'NFLX', 'NVDA', 'NXPI', 'ORLY', 'ODFL', 'ON', 'PCAR', 'PANW', 'PAYX', 'PYPL', 'PDD', 'PEP', 'QCOM', 'REGN', 'ROP', 'ROST', 'SIRI', 'SBUX', 'SNPS', 'TTWO', 'TMUS', 'TSLA', 'TXN', 'TTD', 'VRSK', 'VRTX', 'WBA', 'WBD', 'WDAY', 'XEL', 'ZS']
101


---
<font color=green>Q3: (3 Marks)</font>
<br><font color='green'>
Using **[yfinance](https://pypi.org/project/yfinance/)** library, write a Python script that accepts a list of stock ticker symbols. For each symbol, download the adjusted closing price data, store it in a dictionary with the ticker symbol as the key, and then convert the final dictionary into a Pandas DataFrame. Handle any errors encountered during data retrieval by printing a message indicating which symbol failed
</font>
---

In [3]:
#Q3

import yfinance as yf
import pandas as pd

def download_adjusted_close(tickers):
    data_dict = {}
    
    for ticker in tickers:
        try:
            data = yf.download(ticker, progress=False) 
            data_dict[ticker] = data['Adj Close']
        except Exception as e:
            print(f"Failed to retrieve data for {ticker}: {e}")
    
    nasdaq100 = pd.DataFrame(data_dict)
    
    return nasdaq100

tickers = tickers_list
nasdaq100 = download_adjusted_close(tickers)

print(nasdaq100)

                  ADBE         ADP        ABNB       GOOGL        GOOG  \
Date                                                                     
1962-01-02         NaN         NaN         NaN         NaN         NaN   
1962-01-03         NaN         NaN         NaN         NaN         NaN   
1962-01-04         NaN         NaN         NaN         NaN         NaN   
1962-01-05         NaN         NaN         NaN         NaN         NaN   
1962-01-08         NaN         NaN         NaN         NaN         NaN   
...                ...         ...         ...         ...         ...   
2024-05-31  444.760010  244.919998  144.929993  172.500000  173.960007   
2024-06-03  439.019989  244.020004  146.250000  173.169998  174.419998   
2024-06-04  448.369995  245.669998  147.080002  173.789993  175.130005   
2024-06-05  455.799988  245.779999  145.779999  175.410004  177.070007   
2024-06-06  464.440002  247.634995  147.639999  176.440598  177.940002   

                  AMZN         AMD   

---
<font color=green>Q4: (3 Marks)</font>
<br><font color='green'>
Write a Python script to analyze stock data stored in a dictionary `stock_data` (where each key is a stock ticker symbol, and each value is a Pandas Series of adjusted closing prices). The script should:
1. Convert the dictionary into a DataFrame.
2. Calculate the daily returns for each stock.
3. Identify columns (ticker symbols) with at least 2000 non-NaN values in their daily returns.
4. Create a new DataFrame that only includes these filtered ticker symbols.
5. Remove any remaining rows with NaN values in this new DataFrame.
</font>

---

In [4]:
#Q4

# Find the daily percentage change
stock_data = nasdaq100.pct_change()
#Filter stocks that have more than 2000 NA daily values
valid_columns = stock_data.columns[stock_data.count() >= 2000]
filtered_nasdaq100 = stock_data[valid_columns]
#Drop rows containing NA values
filtered_nasdaq100 = filtered_nasdaq100.dropna()

#Print the filtered DataFrame
print(filtered_nasdaq100)

                ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2015-12-10 -0.006699  0.006004 -0.003292 -0.002861 -0.003715  0.042553   
2015-12-11  0.027653 -0.024924 -0.012657 -0.014130 -0.033473 -0.036735   
2015-12-14  0.020127  0.015240  0.016151  0.012045  0.027744 -0.008475   
2015-12-15  0.008149  0.010284 -0.003213 -0.005844  0.001110  0.008547   
2015-12-16  0.016380  0.008541  0.021708  0.019761  0.026008  0.076271   
...              ...       ...       ...       ...       ...       ...   
2024-05-31 -0.002489  0.016645  0.002266  0.002305 -0.016061  0.000900   
2024-06-03 -0.012906 -0.003675  0.003884  0.002644  0.010768 -0.020072   
2024-06-04  0.021297  0.006762  0.003580  0.004071  0.005607 -0.021767   
2024-06-05  0.016571  0.000448  0.009322  0.011077  0.010817  0.038627   
2024-06-06  0.018956  0.007547  0.005875  0.004913  0.010647 -0.001685   

                 AEP      AMGN       

---
<font color=green>Q5: (1 Mark)</font>
<br><font color='green'>
Download the dataset named `df_filtered_nasdaq_100` from the GitHub repository of the course.
</font>

---

In [5]:
#Q5

#Get file from github page
df_filtered_nasdaq_100 = pd.read_csv("https://raw.githubusercontent.com/Jandsy/ml_finance_imperial/main/Coursework/df_filtered_nasdaq_100.csv",index_col=0, parse_dates=True)

print(df_filtered_nasdaq_100)

                ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2015-12-10 -0.006699  0.006004 -0.003292 -0.002861 -0.003715  0.042553   
2015-12-11  0.027653 -0.024924 -0.012657 -0.014130 -0.033473 -0.036735   
2015-12-14  0.020127  0.015241  0.016151  0.012045  0.027744 -0.008475   
2015-12-15  0.008149  0.010283 -0.003213 -0.005844  0.001110  0.008547   
2015-12-16  0.016380  0.008541  0.021708  0.019761  0.026008  0.076271   
...              ...       ...       ...       ...       ...       ...   
2024-05-06  0.015241  0.003514  0.005142  0.004971  0.013372  0.034396   
2024-05-07 -0.002674  0.009805  0.018739  0.018548  0.000318 -0.008666   
2024-05-08 -0.008471 -0.008894 -0.010920 -0.010521 -0.004026 -0.005245   
2024-05-09 -0.011166  0.009097  0.003424  0.002454  0.007979 -0.008007   
2024-05-10 -0.000746  0.006975 -0.007708 -0.007518 -0.010660 -0.003084   

                 AEP      AMGN       

---
<font color=green>Q6: (3 Marks) </font>
<br><font color='green'>
Conduct an in-depth analysis of the `df_filtered_nasdaq_100` dataset from GitHub. Answer the following questions:
- Which stock had the best performance over the entire period?
- What is the average daily return of 'AAPL'?
- What is the worst daily return? Provide the stock name and the date it occurred.
</font>

---

In [6]:
#Q6

# Calculate cumulative returns for each stock
cumulative_returns = (1 + df_filtered_nasdaq_100).cumprod() 

# Identify the stock with the highest cumulative return
best_performing_stock = cumulative_returns.iloc[-1].idxmax()

# Calculate the average daily return of 'AAPL'
average_daily_return_aapl = df_filtered_nasdaq_100['AAPL'].mean()

# Find the worst daily return and identify the stock name and the date it occurred
daily_returns = df_filtered_nasdaq_100
worst_daily_return_stock = daily_returns.min().idxmin()
worst_daily_return_value = daily_returns[worst_daily_return_stock].min()
worst_daily_return_date = daily_returns[worst_daily_return_stock].idxmin()

# Print the results
print("1. Best performing stock over the entire period:", best_performing_stock)
print("2. Average daily return of 'AAPL':", average_daily_return_aapl)
print("3. Worst daily return occurred on", worst_daily_return_date, "for the stock", worst_daily_return_stock, "with a return of", worst_daily_return_value)

1. Best performing stock over the entire period: NVDA
2. Average daily return of 'AAPL': 0.0010849409515136207
3. Worst daily return occurred on 2020-03-09 00:00:00 for the stock FANG with a return of -0.4464579394678258


# Fama French Analysis

The Fama-French five-factor model is an extension of the classic three-factor model used in finance to describe stock returns. It is designed to better capture the risk associated with stocks and explain differences in returns. This model includes the following factors:

1. **Market Risk (MKT)**: The excess return of the market over the risk-free rate. It captures the overall market's premium.
2. **Size (SMB, "Small Minus Big")**: The performance of small-cap stocks relative to large-cap stocks.
3. **Value (HML, "High Minus Low")**: The performance of stocks with high book-to-market values relative to those with low book-to-market values.
4. **Profitability (RMW, "Robust Minus Weak")**: The difference in returns between companies with robust (high) and weak (low) profitability.
5. **Investment (CMA, "Conservative Minus Aggressive")**: The difference in returns between companies that invest conservatively and those that invest aggressively.

## Additional Factor

6. **Momentum (MOM)**: This factor represents the tendency of stocks that have performed well in the past to continue performing well, and the reverse for stocks that have performed poorly.

### Mathematical Representation

The return of a stock $R_i^t$ at time $t$ can be modeled as follows :

$$
R_i^t - R_f^t = \alpha_i^t + \beta_{i,MKT}^t(R_M^t - R_f^t) + \beta_{i,SMB}^t \cdot SMB^t + \beta_{i,HML}^t \cdot HML^t + \beta_{i,RMW}^t \cdot RMW^t + \beta_{i,CMA}^t \cdot CMA^t + \beta_{i,MOM}^t \cdot MOM^t + \epsilon_i^t
$$

Where:
- $ R_i^t $ is the return of stock $i$ at time $t$
- $R_f^t $is the risk-free rate at time $t$
- $ R_M^t $ is the market return at time $t$
- $\alpha_i^t $ is the abnormal return or alpha of stock $ i $ at time $t$
- $\beta^t $ coefficients represent the sensitivity of the stock returns to each factor at time $t$
- $\epsilon_i^t $ is the error term or idiosyncratic risk unique to stock $ i $ at time $t$

This model is particularly useful for identifying which factors significantly impact stock returns and for constructing a diversified portfolio that is optimized for given risk preferences.




---
<font color=green>Q7: (1 Mark) </font>
<br><font color='green'>
Download the `fama_french_dataset` from the course's GitHub account.
</font>

---

In [7]:
#Q7

#Get the data from the github page
fama_french_data = pd.read_csv("https://github.com/Jandsy/ml_finance_imperial/raw/main/Coursework/fama_french_dataset.csv")

# Rename the first column to 'Date'
fama_french_data = fama_french_data.rename(columns={fama_french_data.columns[0]: 'Date'})

# Convert 'Date' column to datetime
fama_french_data['Date'] = pd.to_datetime(fama_french_data['Date'])

"""Filter data between the dates we want. If we want to filter by the papers' dates we can change the values accordingly"""  
start_date = '1963-01-07'
end_date = '2024-03-28'
filtered_date = (fama_french_data['Date'] >= start_date) & (fama_french_data['Date'] <= end_date)
fama_french_data = fama_french_data.loc[filtered_date]

# Set 'Date' as the index
fama_french_data.set_index('Date', inplace=True)

print(fama_french_data)

            Mkt-RF   SMB   HML   RMW   CMA     RF   Mom
Date                                                   
1963-07-01   -0.67  0.02 -0.35  0.03  0.13  0.012 -0.21
1963-07-02    0.79 -0.28  0.28 -0.08 -0.21  0.012  0.42
1963-07-03    0.63 -0.18 -0.10  0.13 -0.25  0.012  0.41
1963-07-05    0.40  0.09 -0.28  0.07 -0.30  0.012  0.07
1963-07-08   -0.63  0.07 -0.20 -0.27  0.06  0.012 -0.45
...            ...   ...   ...   ...   ...    ...   ...
2024-03-22   -0.23 -0.98 -0.53  0.29 -0.37  0.021  0.43
2024-03-25   -0.26 -0.10  0.88 -0.22 -0.17  0.021 -0.34
2024-03-26   -0.26  0.10 -0.13 -0.50  0.23  0.021  0.09
2024-03-27    0.88  1.29  0.91 -0.14  0.58  0.021 -1.34
2024-03-28    0.10  0.45  0.48 -0.07  0.09  0.021 -0.44

[15290 rows x 7 columns]


---
<font color=green>Q8: (5 Marks)</font>
<br><font color='green'>

Write a Python function called `get_sub_df_ticker(ticker, date, df_filtered, length_history)` that extracts a historical sub-dataframe for a given `ticker` from `df_filtered`. The function should use `length_history` to determine the number of trading days to include, ending at the specified `date`. Return the sub-dataframe for the specified `ticker`.
</font>

---


In [8]:
#Q8

def get_sub_df_ticker(ticker, date, df_filtered_nasdaq_100, length_history):
    
    date = pd.to_datetime(date)
    
    ticker_data = df_filtered_nasdaq_100[[ticker]]

    end_index = ticker_data.index.get_loc(date)

    start_index = max(0, end_index - length_history + 1)
   
    sub_df = ticker_data.iloc[start_index:end_index + 1]
    
    return sub_df

#Example of function usage:
ticker = 'AMD'
date = '2021-01-04'
length_history =252

sub_df = get_sub_df_ticker(ticker, date, df_filtered_nasdaq_100, length_history)

#Display the DataFrame
print(sub_df)

                 AMD
Date                
2020-01-06 -0.004321
2020-01-07 -0.002893
2020-01-08 -0.008705
2020-01-09  0.023834
2020-01-10 -0.016337
...              ...
2020-12-28 -0.002287
2020-12-29 -0.010699
2020-12-30  0.018429
2020-12-31 -0.006285
2021-01-04  0.006433

[252 rows x 1 columns]


---
<font color=green>Q9: (4 Marks)</font>
<br><font color='green'>
Create a Python function named `df_ticker_with_fama_french(ticker, date, df_filtered, length_history, fama_french_data)` that uses `get_sub_df_ticker` to extract historical data for a specific `ticker`. Incorporate the Fama-French factors from `fama_french_data` into the extracted sub-dataframe. Adjust the ticker's returns by subtracting the risk-free rate ('RF') and add other relevant Fama-French factors ('Mkt-RF', 'SMB', 'HML', 'RMW', 'CMA', and 'Mom'). Return the resulting sub-dataframe.
</font>

---

In [9]:
#Q9 

def df_ticker_with_fama_french(ticker, date, df_filtered_nasdaq_100, length_history, fama_french_data):
    
    sub_df = get_sub_df_ticker(ticker, date, df_filtered_nasdaq_100, length_history)
    
    # Join stock data with Fama-French factors
    combined_df = sub_df.join(fama_french_data, how='inner') 
    
    # Calculate Excess Return
    combined_df['Excess Return'] = combined_df[ticker] - combined_df['RF']
    
    return combined_df


combined_df = df_ticker_with_fama_french(ticker, date, df_filtered_nasdaq_100, length_history, fama_french_data)

#Display the DataFrame
print(combined_df)


                 AMD  Mkt-RF   SMB   HML   RMW   CMA     RF   Mom  \
Date                                                                
2020-01-06 -0.004321    0.36 -0.20 -0.55 -0.17 -0.26  0.006 -0.69   
2020-01-07 -0.002893   -0.19 -0.03 -0.25 -0.12 -0.25  0.006  0.01   
2020-01-08 -0.008705    0.47 -0.17 -0.64 -0.19 -0.17  0.006  0.92   
2020-01-09  0.023834    0.65 -0.71 -0.48 -0.14  0.04  0.006  0.73   
2020-01-10 -0.016337   -0.34 -0.27 -0.33  0.04 -0.08  0.006  0.18   
...              ...     ...   ...   ...   ...   ...    ...   ...   
2020-12-28 -0.002287    0.46 -0.68  0.36  1.39  0.46  0.000 -0.47   
2020-12-29 -0.010699   -0.40 -1.42  0.24  0.78 -0.29  0.000 -0.41   
2020-12-30  0.018429    0.27  1.06  0.03 -0.65 -0.05  0.000 -0.27   
2020-12-31 -0.006285    0.39 -0.71  0.41  0.57 -0.24  0.000 -0.59   
2021-01-04  0.006433   -1.41  0.16  0.58 -0.64  0.10  0.000 -0.04   

            Excess Return  
Date                       
2020-01-06      -0.010321  
2020-01-07      -0

---
<font color=green>Q10: (5 Marks) </font>
<br><font color='green'>
Write a Python function named `extract_beta_fama_french` to perform a rolling regression analysis for a given stock at a specific time point using the Fama-French model. The function should accept the following parameters:

- `ticker`: A string indicating the stock symbol.
- `date`: A string specifying the date for the analysis.
- `length_history`: An integer representing the number of days of historical data to include.
- `df_filtered`: A pandas DataFrame (assumed to be derived from question 5) containing filtered stock data.
- `fama_french_data`: A pandas DataFrame (assumed to be from question 7) that includes Fama-French factors.

Utilize the `statsmodels.api` library to conduct the regression.
</font>

---

In [10]:
#Q10

import statsmodels.api as sm

def extract_beta_fama_french(ticker, date, length_history, df_filtered_nasdaq_100, fama_french_data):
    
    # Ensure the date is in datetime format
    date = pd.to_datetime(date)
    
    # Get function from previous question and specify window size
    combined_df = df_ticker_with_fama_french(ticker, date, df_filtered_nasdaq_100, length_history, fama_french_data)
    window_size = 60

    rolling_betas = []
    rolling_residuals = []

    dates = combined_df.index[window_size:]
    
    # This is used to find R Squared of the model needed for question 19
    last_model = None


    for ending_date in dates:
        try:
            #Starting date for rolling window
            starting_date = ending_date - pd.DateOffset(days=window_size)
            #Use combined dataframe to extract relevant data to window df
            window_df = combined_df.loc[starting_date:ending_date]
            y = window_df['Excess Return']
            X = window_df[['Mkt-RF', 'SMB', 'HML', 'RMW', 'CMA', 'Mom']]
            X = sm.add_constant(X)
            model = sm.OLS(y, X).fit()
            #Extract betas and residuals
            betas = model.params
            residuals = model.resid
            betas['Date'] = ending_date
            rolling_betas.append(betas)
            
            
            last_model = model

            #Store the mean residual for the current window's ending_date
            rolling_residuals.append({'Date': ending_date, 'Residual': residuals.mean()})
            
        except Exception as e:
            print(f"Failed to run regression for {ending_date}: {e}")
            continue

    # Combine betas into a DataFrame
    betas_df = pd.DataFrame(rolling_betas).set_index('Date')

    # Combine residuals into a DataFrame
    residuals_df = pd.DataFrame(rolling_residuals).set_index('Date')
    
    #Calculate R^2
    r_squared = last_model.rsquared if last_model else None

    return betas_df, residuals_df, model.summary(),r_squared

ff_results = extract_beta_fama_french(ticker, date, length_history, df_filtered_nasdaq_100, fama_french_data)

---
<font color=green>Q11: (2 Marks) </font>
<br><font color='green'>
Apply the `extract_beta_fama_french` function to the stock symbol 'AAPL' for the date '2024-03-28', using a historical data length of 252 days. Ensure that the `df_filtered` and `fama_french_data` DataFrames are correctly prepared and available in your environment before executing this function. The parameters for the function call are set as follows:

- **Ticker**: 'AAPL'
- **Date**: '2024-03-28'
- **Length of History**: 252 days
</font>

---



In [11]:
#Q11

ticker = 'AAPL'
date = '2024-03-28'
length_history = 252

#Call the function we created before
ff_results_apple = extract_beta_fama_french(ticker, date, length_history, df_filtered_nasdaq_100, fama_french_data)

#Print the results for AAPL
print(ff_results_apple[0])  # Betas DataFrame
print(ff_results_apple[1])  # Residuals DataFrame

               const    Mkt-RF       SMB       HML       RMW       CMA  \
Date                                                                     
2023-06-26 -0.016966  0.010593 -0.001548 -0.003647  0.004572 -0.001838   
2023-06-27 -0.017074  0.010140 -0.001437 -0.003380  0.004252 -0.002091   
2023-06-28 -0.016942  0.010157 -0.001213 -0.003245  0.003928 -0.002249   
2023-06-29 -0.017018  0.010088 -0.001325 -0.003162  0.003801 -0.002438   
2023-06-30 -0.016837  0.010572 -0.001556 -0.003048  0.003414 -0.002081   
...              ...       ...       ...       ...       ...       ...   
2024-03-22 -0.024051  0.006684  0.002445 -0.011018  0.005096  0.003073   
2024-03-25 -0.024377  0.006824  0.000997 -0.010806  0.004160  0.002315   
2024-03-26 -0.024405  0.006677  0.001091 -0.010902  0.004264  0.001898   
2024-03-27 -0.023653  0.007882  0.001642 -0.011233  0.005588  0.002879   
2024-03-28 -0.023836  0.007968  0.001566 -0.011303  0.005548  0.003230   

                 Mom  
Date          

---
<font color=green>Q12: (2 Marks)</font>
<br><font color='green'>
Once the `extract_beta_fama_french` function has been applied to 'AAPL' with the specified parameters, the next step is to analyze the regression summary to identify which Fama-French factor explains the most variance in 'AAPL' returns during the specified period.

Follow these steps to perform the analysis:

1. **Review the Summary**: Examine the regression output, focusing on the coefficients and their statistical significance (p-values).
2. **Identify Key Factor**: Determine which factor has the highest absolute coefficient value and is statistically significant (typically p < 0.05). This factor can be considered as having the strongest influence on 'AAPL' returns for the period.

</font>

---

In [12]:
#Q12

ff_results_apple = extract_beta_fama_french(ticker, date, length_history, df_filtered_nasdaq_100, fama_french_data)

#Print OLS Regression Results
print(ff_results_apple[2])

                            OLS Regression Results                            
Dep. Variable:          Excess Return   R-squared:                       0.385
Model:                            OLS   Adj. R-squared:                  0.283
Method:                 Least Squares   F-statistic:                     3.760
Date:                Thu, 06 Jun 2024   Prob (F-statistic):            0.00524
Time:                        17:50:55   Log-Likelihood:                 138.32
No. Observations:                  43   AIC:                            -262.6
Df Residuals:                      36   BIC:                            -250.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0238      0.002    -13.392      0.0

The factor that is statistically significant (P-Value < 0.05) and with the highest coefficient value is the Mkt-RF factor, representing the excess market return over the risk free rate. This is the factor with the highest influence for AAPL

# PCA Analysis


In literature, another method exists for extracting residuals for each stock, utilizing the PCA approach to identify hidden factors in the data. Let's describe this method.

The return of a stock $R_i^t$ at time $t$ can be modeled as follows :

$$
R_i^t  = \sum_{j=1}^m\beta_{i,j}^t F_j^t  + \epsilon_i^t
$$

Where:
- $ R_i^t $ is the return of stock $i$ at time $t$
- $m$ is the number of factors selected from PCA
-  $ F_j^t $ is the $j$-th hidden factor constructed from PCA at time $t$
- $\beta_{i,j}^t $ are the coefficients representing the sensitivity of the stock returns to each hidden factor.
- $\epsilon_i^t $  is the residual term for stock $i$ at time $t$, representing the portion of the return not explained by the PCA factors.

### Representation of Stock Return Data

Consider the return data for $N$ stocks over $T$ periods, represented by the matrix $R$ of size $T \times N$:

$$
R = \left[
\begin{array}{cccc}
R_1^T & R_2^T & \cdots & R_N^T \\
R_1^{T-1} & R_2^{T-1} & \cdots & R_N^{T-1} \\
\vdots & \vdots & \ddots & \vdots \\
R_1^1 & R_2^1 & \cdots & R_N^1 \\
\end{array}
\right]
$$

Each element $R_i^k$ of the matrix represents the return of stock $i$ at time $k$ and is defined as:

$$
R_i^k = \frac{S_{i,k} - S_{i, k-1}}{S_{i, k-1}}, \quad k=1,\cdots, T, \quad i=1,\cdots,N
$$

where $S_{i,k}$ denotes the adjusted close price of stock $i$ at time $k$.

### Standardization of Returns

To adjust for varying volatilities across stocks, we standardize the returns as follows:

$$
Z_i^t = \frac{R_i^t - \mu_i}{\sigma_i}
$$

where $\mu_i$ and $\sigma_i$ are the mean and standard deviation of returns for stock $i$ over the period $[t-T, t]$, respectively.

### Empirical Correlation Matrix

The empirical correlation matrix $C$ is computed from the standardized returns:

$$
C = \frac{1}{T-1} Z^T Z
$$

where $Z^T$ is the transpose of matrix $Z$.

### Singular Value Decomposition (SVD)

We apply Singular Value Decomposition to the correlation matrix $C$:

$$
C = U \Sigma V^T
$$

Here, $U$ and $V$ are orthogonal matrices representing the left and right singular vectors, respectively, and $\Sigma$ is a diagonal matrix containing the singular values, which are the square roots of the eigenvalues.

### Construction of Hidden Factors

For each of the top $m$ components, we construct the selected hidden factors as follows:

$$
F_j^t = \sum_{i=1}^N \frac{\lambda_{i,j}}{\sigma_i} R_i^t
$$

where $\lambda_{i,j}$ is the $i$-th component of the $j$-th eigenvector (ranked by eigenvalue magnitude).


---
<font color=green>Q13 (3 Marks):

For the specified period from March 29, 2023 ('2023-03-29'), to March 28, 2024 ('2024-03-28'), generate the matrix $Z$ by standardizing the stock returns using the DataFrame `df_filtered_new`
</font>

---


In [13]:
#Q13

# Define the start and end dates for the period
start_date = '2023-03-29'
end_date = '2024-03-28'

# Filter the DataFrame for the specified period
df_filtered_new = df_filtered_nasdaq_100[start_date:end_date]

# Calculate the mean and standard deviation for each stock over the specified period
mean = df_filtered_new.mean()
std_dev = df_filtered_new.std()

# Standardize the returns
Z = (df_filtered_new - mean) / std_dev

# Print the resulting matrix Z
print(Z)

                ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2023-03-29  0.658925  2.189521  0.106285  0.207821  1.503570  0.444433   
2023-03-30  0.273051 -0.221306 -0.389329 -0.434766  0.787034  0.526996   
2023-03-31  0.360570  1.136308  1.540732  1.439604  0.531735 -0.056439   
2023-04-03 -0.713053 -2.259591  0.252735  0.407399 -0.591892 -0.600163   
2023-04-04  0.560730 -1.137430  0.099652  0.013877  0.658632 -0.342217   
...              ...       ...       ...       ...       ...       ...   
2024-03-22 -1.146803 -0.516681  1.151431  1.085070  0.074915  0.081847   
2024-03-25  0.659349 -1.221008 -0.372489 -0.341074  0.109668 -0.292706   
2024-03-26 -0.032708  0.234343  0.131651  0.109342 -0.556130 -0.244717   
2024-03-27 -0.363721  1.052056 -0.024166 -0.010591  0.315892  0.224881   
2024-03-28 -0.048375  0.411934 -0.078409  0.019962  0.022729  0.067776   

                 AEP      AMGN       

---
<font color=green>Q14: (1 Mark) </font>
<br><font color='green'>
Download the `Z_matrix` matrix from the course's GitHub account.
</font>

---

In [14]:
#Q14

Z_matrix =pd.read_csv("https://raw.githubusercontent.com/Jandsy/ml_finance_imperial/main/Coursework/Z_matrix.csv",index_col=0, parse_dates=True)

#Display the Z_matrix from the Github account
print(Z_matrix)

                ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2023-03-29  0.658925  2.189521  0.106285  0.207821  1.503570  0.444433   
2023-03-30  0.273051 -0.221306 -0.389329 -0.434766  0.787034  0.526996   
2023-03-31  0.360570  1.136308  1.540732  1.439604  0.531735 -0.056439   
2023-04-03 -0.713053 -2.259591  0.252735  0.407399 -0.591892 -0.600163   
2023-04-04  0.560730 -1.137430  0.099652  0.013877  0.658632 -0.342217   
...              ...       ...       ...       ...       ...       ...   
2024-03-22 -1.146803 -0.516681  1.151431  1.085070  0.074915  0.081847   
2024-03-25  0.659349 -1.221008 -0.372489 -0.341074  0.109668 -0.292706   
2024-03-26 -0.032708  0.234343  0.131651  0.109342 -0.556130 -0.244717   
2024-03-27 -0.363721  1.052056 -0.024166 -0.010591  0.315892  0.224881   
2024-03-28 -0.048375  0.411934 -0.078409  0.019962  0.022729  0.067776   

                 AEP      AMGN       

---
<font color=green>Q15: (3 Marks) </font>
<br><font color='green'>
For the specified period from March 29, 2023 ('2023-03-29'), to March 28, 2024 ('2024-03-28'), compute the correlation matrix
$C$ using the matrix `Z_matrix`.
</font>

---

In [15]:
#Q15

#Calculate Transpose of matrix Z
Z_transpose = Z_matrix.T

#Calculate Correlation Matrix
correlation_matrix = (1/(length_history-1)) * Z_transpose.dot(Z_matrix)

# Display the correlation matrix
print(correlation_matrix)

           ADBE       ADP     GOOGL      GOOG      AMZN       AMD       AEP  \
ADBE   1.000000  0.218513  0.397890  0.400601  0.463488  0.444032 -0.035967   
ADP    0.218513  1.000000  0.294213  0.298841  0.168206  0.045884  0.228457   
GOOGL  0.397890  0.294213  1.000000  0.997415  0.521199  0.371105 -0.006803   
GOOG   0.400601  0.298841  0.997415  1.000000  0.525626  0.371568 -0.004037   
AMZN   0.463488  0.168206  0.521199  0.525626  1.000000  0.463049 -0.010849   
...         ...       ...       ...       ...       ...       ...       ...   
VRTX   0.164760  0.176771  0.142447  0.146190  0.104962  0.039540  0.239861   
WBA    0.033955  0.142369  0.052710  0.060822  0.017926  0.002629  0.309717   
WBD    0.099841  0.243986  0.042072  0.045516  0.162937  0.092733  0.325463   
WDAY   0.418110  0.320836  0.289137  0.293575  0.403757  0.334587  0.017659   
XEL    0.019105  0.164682  0.025701  0.026392 -0.058870 -0.214673  0.617318   

           AMGN       ADI      ANSS  ...      TTWO 

---
<font color=green>Q16: (2 Marks) </font>
<br><font color='green'>
Refind the correlation matrix from the from March 29, 2023 ('2023-03-29'), to March 28, 2024 ('2024-03-28') using pandas correlation matrix method.
</font>

---

In [16]:
#Q16

#Use Pandas Method
correlation_matrix_pandas = Z_matrix.corr()

# Display the correlation matrix 
print(correlation_matrix_pandas)

           ADBE       ADP     GOOGL      GOOG      AMZN       AMD       AEP  \
ADBE   1.000000  0.218513  0.397890  0.400601  0.463488  0.444032 -0.035967   
ADP    0.218513  1.000000  0.294213  0.298841  0.168206  0.045884  0.228457   
GOOGL  0.397890  0.294213  1.000000  0.997415  0.521199  0.371105 -0.006803   
GOOG   0.400601  0.298841  0.997415  1.000000  0.525626  0.371568 -0.004037   
AMZN   0.463488  0.168206  0.521199  0.525626  1.000000  0.463049 -0.010849   
...         ...       ...       ...       ...       ...       ...       ...   
VRTX   0.164760  0.176771  0.142447  0.146190  0.104962  0.039540  0.239861   
WBA    0.033955  0.142369  0.052710  0.060822  0.017926  0.002629  0.309717   
WBD    0.099841  0.243986  0.042072  0.045516  0.162937  0.092733  0.325463   
WDAY   0.418110  0.320836  0.289137  0.293575  0.403757  0.334587  0.017659   
XEL    0.019105  0.164682  0.025701  0.026392 -0.058870 -0.214673  0.617318   

           AMGN       ADI      ANSS  ...      TTWO 

---
<font color=green>Q17: (7 Marks) </font>
<br><font color='green'>
Conduct Singular Value Decomposition on the correlation matrix $C$. Follow these steps:


1.   **Perform SVD**: Decompose the matrix $C$ into its singular values and vectors.
2.   **Rank Eigenvalues**: Sort the resulting singular values (often squared to compare to eigenvalues) in descending order.
3. **Select Components**: Extract the first 20 components based on the largest singular values.
4. **Variance Explained**: Print the variance explained by the first 20 Components and dimensions of differents matrix that you created.

</font>

---

In [17]:
#Q17

import numpy as np

# Perform SVD
SVD = np.linalg.svd(correlation_matrix_pandas)

# Calculate and Rank Eigenvalues
eigenvalues = SVD[1]**2
sorted_eigenvalues = np.argsort(eigenvalues)[::-1]

# Select the first 20 eigenvalues
top_20_indices = sorted_eigenvalues[:20]
top_20_eigenvalues = eigenvalues[top_20_indices]

# Find variance of model explained by the first 20 eigenvalues
explained_variance = np.sum(top_20_eigenvalues) / np.sum(eigenvalues)
explained_variance_pct = top_20_eigenvalues / np.sum(eigenvalues)

# Print dimensions of matrices
print("\nDimensions of the matrices:")
print("U matrix:", SVD[0].shape)
print("S (eigenvalues):", SVD[1].shape)
print("Vt matrix:", SVD[2].shape)

# Print the top 20 eigenvalues
print("\nTop 20 components:",top_20_eigenvalues)

print ('\nExplained Variance percentage: ', explained_variance_pct)


# Print the variance explained by the first 20 components
print("\nVariance explained by the first 20 components:",explained_variance*100,"%")


Dimensions of the matrices:
U matrix: (89, 89)
S (eigenvalues): (89,)
Vt matrix: (89, 89)

Top 20 components: [489.80714433  53.88241254  15.34254661   7.62427837   6.63908465
   4.19347111   3.45696857   3.10609749   3.03317018   2.60015846
   2.30307941   2.09724098   1.96779575   1.75439203   1.56970662
   1.42293776   1.37379448   1.31867962   1.1929447    1.15649058]

Explained Variance percentage:  [0.78617886 0.0864855  0.02462599 0.01223756 0.01065625 0.00673085
 0.00554871 0.00498553 0.00486848 0.00417346 0.00369662 0.00336624
 0.00315847 0.00281594 0.0025195  0.00228393 0.00220505 0.00211658
 0.00191477 0.00185626]

Variance explained by the first 20 components: 97.24245358638191 %


---
<font color=green>Q18: (6 Marks) </font>
<br><font color='green'>
Extract the 20 hidden factors in a matrix F. Check that shape of F is $(252,20)$
</font>

</font>

---

In [18]:
#Q18

# We extract the top 20 components from the V transpose(right singular) matrix of SVD:
top_20_eigenvectors = SVD[2][top_20_indices, :].T

#Get a new dataframe with the standard deviations of returns for each stock:
sigma_returns = df_filtered_new.std()

#Create an initial F matrix
F =np.zeros((length_history,20))

#Use the provided formula to calculate the F matrix top 20 hidden factors
for j in range(20):
    F[:, j] = np.dot(df_filtered_new.values, top_20_eigenvectors[:, j] / sigma_returns.values)

# Double check the shape of the F Matrix
print("Shape of F Matrix:", F.shape)  # Must be (252, 20)
print(F)

Shape of F Matrix: (252, 20)
[[-10.42428276  -1.13807494  -1.48029262 ...   1.23955006   2.27676229
   -0.13751588]
 [ -4.05837617   0.78059648  -0.86611386 ...  -0.36262048  -0.97011478
   -0.91996453]
 [ -7.95739081  -2.27970763   1.48389609 ...   0.05732551  -0.36311986
   -0.11338118]
 ...
 [  0.88061846   0.13735036   1.12187707 ...  -0.17879579   0.09227425
    0.16559638]
 [ -5.48903391  -4.97318897  -3.13593018 ...  -1.29272596  -0.13181172
    0.74442252]
 [ -0.52987874  -1.28782948  -0.79107298 ...  -0.58448879   0.61863981
    0.69301072]]


---
<font color=green>Q19: (3 Marks) </font>
<br><font color='green'>
Perform the Regression Analysis of 'AAPL' for the date '2024-03-28', using a historical data length of 252 days using previous $F$ Matrix. Compare the R-squared from the ones obtained at Q11.
</font>

</font>

---

In [19]:
#Q19

# Utilize relevant data for AAPL
aapl_returns = df_filtered_new['AAPL']
F_regression = pd.DataFrame(F, index=aapl_returns.index)

# Perform the regression
X = F_regression
X = sm.add_constant(X)
y = aapl_returns
model = sm.OLS(y, X).fit()

# Print the summary of the regression model
print(model.summary())

# Compare R-squared from the previous Fama-French regression in Q12
print(f"R-squared from PCA model:", model.rsquared*100,"%")
print(f"R-squared from Fama-French model:", ff_results_apple[3]*100,"%")

                            OLS Regression Results                            
Dep. Variable:                   AAPL   R-squared:                       0.615
Model:                            OLS   Adj. R-squared:                  0.582
Method:                 Least Squares   F-statistic:                     18.48
Date:                Thu, 06 Jun 2024   Prob (F-statistic):           2.29e-37
Time:                        17:50:55   Log-Likelihood:                 873.13
No. Observations:                 252   AIC:                            -1704.
Df Residuals:                     231   BIC:                            -1630.
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0005      0.001     -0.874      0.3

# Ornstein Uhlenbeck

The Ornstein-Uhlenbeck process is defined by the following stochastic differential equation (SDE):

$$ dX_t = \theta (\mu - X_t) dt + \sigma dW_t $$

where:

- **$ X_t $**: The value of the process at time $ t $.
- **$ \mu $**: The long-term mean (equilibrium level) to which the process reverts.
- **$ \theta $**: The speed of reversion or the rate at which the process returns to the mean.
- **$ \sigma $**: The volatility (standard deviation), representing the magnitude of random fluctuations.
- **$ W_t $**: A Wiener process or Brownian motion that adds stochastic (random) noise.

This equation describes a process where the variable $ X_t $ moves towards the mean $ \mu $ at a rate determined by $ \theta $, with random noise added by $ \sigma dW_t $.

---
<font color=green>Q20: (3 Marks) </font>
<br><font color='green'>
In the context of mean reversion, which quantity should be modeled using an Ornstein-Uhlenbeck process?
</font>

---

The quantity that should be modeled using and Ornstein-Uhlenbeck process is the residuals of the time series for several reasons. First, the solution to the model is in a closed form, making it tractable without needing to apply numerical approximations. Also, since residual are assumed to be stochastic, the model accurately replicates that by assuming that the residuals follow a standard normal distribution under this process. Furthermore, the first and second moments (mean and variance) are defined, and the process is mean reverting, which sets a limit for the drift. All these assumptions and properties make the model favourable for describing the behaviour of the residuals in time series analysis while maintaining simplicity and trackability.

---
<font color=green>Q21: (5 Marks) </font>
<br><font color='green'>
Explain how the parameters $ \theta $ and $ \sigma $ can be determined using the following equations. Also, detail the underlying assumptions:
$$ E[X] = \mu $$
$$ \text{Var}[X] = \frac{\sigma^2}{2\theta} $$
</font>

---

### Determining the parameters $\theta$ and $\sigma$ 

The parameters $\theta$ and $\sigma$ can be determined using maximum likelihood estimation. Knowing that the solution to the stochastic differential equation (SDE) is in continuous time, we need to transition into a discrete time framework. 

After modelling the stock returns, we extract the daily residuals and calculate their first difference as a proxy for $dX_t$. Then we either assume that the residuals follow a certain distribution such as the standard normal distribution, or we perform tests to extract the properties on their distribution. 

We then write the log likelihood of the parameters and maximize that function using argmax:

$$\theta^*, sigma^* = \text{argmax}_{\theta, \sigma} \ell(\theta, \sigma; \{X_{t_i}\})$$

where $\ell$ is the log-likelihood function estimated based on the density function of the residuals. This maximization can be achieved using numerical methods or using gradient descent. Finally, the obtained $\theta^*$ and $\sigma^*$ are used as the parameter in the Ornstein-Uhlenbeck process to model the residual time series. 



### Analysis of the Model

To understand the properties of the model, we must first solve the stochatic differential equation, then we must compute the unconditional expectation and variance of $X_t$.

The SDE for the Ornstein-Uhlenbeck process is given by:
$$dX_t = \theta (\mu - X_t) dt + \sigma dW_t$$

$$dX_t = \theta \mu dt - \theta X_t dt + \sigma dW_t$$


Multiply the equation by the Intergrating factor $e^{\theta t}$:

$$e^{\theta t} dX_t = e^{\theta t} \theta \mu dt - e^{\theta t} \theta X_t dt + e^{\theta t} \sigma dW_t$$

Note:
$$\frac{d}{dt}(X_t e^{\theta t}) = e^{\theta t} \frac{dX_t}{dt} + X_t \frac{d}{dt}(e^{\theta t})$$

Thus: 
$$\frac{d}{dt}(X_t e^{\theta t}) = \theta \mu e^{\theta t} + \sigma e^{\theta t} \frac{dW_t}{dt}$$


$$ dX_t e^{\theta t} = \theta \mu e^{\theta t}dt + \sigma e^{\theta t}dW_t$$


Integrate both sides from $t$ to $T$:
$$\int_t^T d(X_s e^{\theta s}) = \int_t^T \theta \mu e^{\theta s} ds +\int_t^T \sigma e^{\theta s} dW_s$$
$$X_T e^{\theta T} - X_t e^{\theta t} = \mu e^{\theta T} - \mu e^{\theta t} + \int_t^T \sigma e^{\theta s} dW_s$$

Isolating $X_t$:
$$X_T = X_t e^{-\theta (T- t)} + \mu (1 - e^{-\theta (T-t)}) + \sigma e^{-\theta T} \int_t^T e^{\theta s} dW_s$$

Let $T=t$ and $t=0$:

$$X_t = X_0 e^{-\theta t} + \mu (1 - e^{-\theta t}) + \sigma e^{-\theta t} \int_0^t e^{\theta s} dW_s$$


### Expectation of $X_t$:

To obtain the expectation of $X_t$, we take the expectation of our solution on both sides:

$$E[X_t] = E[X_0] e^{-\theta t} + \mu (1 - e^{-\theta t}) + 0$$

$$E[X_t] = \mu e^{-\theta t} + \mu - \mu e^{-\theta t}$$

$$E[X_t] = \mu$$

### Explanation: 

$\mu$ is the mean of the model, which is also the unconditional expectation of $X_t$. Looking back at the initial equation describing the Ornstein-Uhlenbeck process, we can realize that this model implies that whenever $X_t$ diverts from the mean $\mu$, it will converge back to $\mu$ at a speed of $\theta$. For that reason, the model is designed such that the unconditional expectation is the mean, highlighting this property of mean reversion. For the case of residual time series, the error follows a brownian motion with a mean-reverting component which indicates that the residual time series converges over time to an expected value equal to its mean, which could be 0 to satisfy the unbias assumption since we are dealing with time series modeling of the residuals and not of stock returns themselves.



### Variance of $X_t$

We now calculate the unconditional variance of $X_t$. Recall the following: 

$$X_t = X_0 e^{-\theta t} + \mu (1 - e^{-\theta t}) + \sigma e^{-\theta t} \int_0^t e^{\theta s} dW_s$$

We know that $V(X_t) = E[(X_t)^2]-(E[X_t])^2$

Hence, we start by calculating $(X_t)^2$:

$$(X_t)^2 = [X_0 e^{-\theta t} + \mu (1 - e^{-\theta t}) + \sigma e^{-\theta t} \int_0^t e^{\theta s} dW_s]\times  [X_0 e^{-\theta t} + \mu (1 - e^{-\theta t}) + \sigma e^{-\theta t} \int_0^t e^{\theta s} dW_s]$$


$$(X_t)^2 = X^2_0 e^{-2\theta t}+ 2X_0\mu e^{-\theta t} (1 - e^{-\theta t})+2X_0\sigma e^{-2\theta t} \int_0^t e^{\theta s} dW_s + u^2 (1-e^{-\theta t})^2 + 2\mu\sigma(1-e^{-\theta t})e^{-\theta t}\int_0^t e^{\theta s} dW_s + \sigma^2 e^{-2\theta t} (\int_0^t e^{\theta s} dW_s)^2$$

Note that according to Ito's multiplication rule, $(dW_s)^2 = ds$. Thus:

$$(\int_0^t e^{\theta s} dW_s)^2 = \int_0^t e^{2\theta s} ds =  \frac{1}{2\theta} (e^{2\theta t}-1)$$

Plug into the equation of $(X_t)^2$:

$$(X_t)^2 = X^2_0 e^{-2\theta t}+ 2X_0\mu e^{-\theta t} (1 - e^{-\theta t})+2X_0\sigma e^{-2\theta t} \int_0^t e^{\theta s} dW_s + u^2 (1-e^{-\theta t})^2 + 2\mu\sigma(1-e^{-\theta t})e^{-\theta t}\int_0^t e^{\theta s} dW_s + \frac{\sigma^2}{2\theta} e^{-2\theta t}   (e^{2\theta t}-1)$$

Note: $\frac{\sigma^2}{2\theta} e^{-2\theta t}(e^{2\theta t}-1) = \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$

Thus: 

$$(X_t)^2 = X^2_0 e^{-2\theta t}+ 2X_0\mu e^{-\theta t} (1 - e^{-\theta t})+ 2X_0\sigma e^{-2\theta t} \int_0^t e^{\theta s} dW_s + u^2 (1-e^{-\theta t})^2 + 2\mu\sigma(1-e^{-\theta t})e^{-\theta t}\int_0^t e^{\theta s} dW_s + \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$$

We now take the expectation of $(X_t)^2$. Note that the expectation of a brownian motion is 0. Also, the expectation of the mean $\mu$ is itself. Thus: 

$$E[X_t]^2 = E[X^2_0] e^{-2\theta t} + E[2X_0]\mu e^{-\theta t}(1 - e^{-\theta t}) + u^2 (1-e^{-\theta t})^2 + \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$$ 

Moreover, we know that $E[X_t] = \mu$. We also set that   $E[X_0] = \mu$. furthermore, since $X_0$ is known, we conclude that  $E[X_0]^2 = \mu^2$. Hence: 

$$E[X_t]^2 = u^2 e^{-2\theta t} + 2\mu^2  e^{-\theta t}(1 - e^{-\theta t}) + u^2 (1-e^{-\theta t})^2 + \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$$ 

$$E[X_t]^2 = u^2 e^{-2\theta t} + 2\mu^2  e^{-\theta t} - 2\mu^2  e^{-2\theta t} + u^2 -2u^2e^{-\theta t}+u^2e^{-2\theta t} + \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$$

All $\mu^2$ terms cancel out except for $\mu^2$:

$$E[X_t]^2 = \mu^2 + \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$$

We now compute the variance: 

$$V(X_t) = E[(X_t)^2]-(E[X_t])^2 = \mu^2 + \frac{\sigma^2}{2\theta}(1-e^{-2\theta t}) - \mu^2$$

$$V(X_t) = \frac{\sigma^2}{2\theta}(1-e^{-2\theta t})$$

Finally, to compute the unconditional variance, we evaluate the limit of $V(X_t)$ as $t$ tends to $\infty$: 

$$V(X) = \lim_{t \to \infty} V_t = \lim_{t \to \infty} \frac{\sigma^2}{2\theta} \left[ 1 - e^{-2\theta t} \right]$$

$$V(X) = \frac{\sigma^2}{2\theta}$$

### Explanation

The variance of $X_t$ in the long term is dependent on two factors: the volatility $\sigma$ and the speed of mean reversion $\theta$. This means that over time we expect $X_t$ to disperse from the mean by $\frac{\sigma^2}{2\theta}$ on average. as the volatility increases or the speed of mean reversion decreases (or both), the variance increases, meaning that the residuals of the time series data will be more dispered and will take a higher range of values away from the mean. This is logical as for high values of $\theta$, the mean reversion component will prevent the residuals from being much larger or much lower from the mean. Also, a greater diffusion coefficient causes the residuals to overcome the mean reversion factor and increase its range of possible values in a stochastic manner before being pulled back towards the mean value. 




---
<font color=green>Q22: (2 Marks) </font>
<br><font color='green'>
Create a function named `extract_s_scores` which computes 's scores' for the last element in a list of floating-point numbers. This function calculates the scores using the following formula $ \text{s scores} = \frac{X_T - \mu}{\sigma} $ where `list_xi` represents a list containing a sequence of floating-point numbers $(X_0, \cdots, X_T)$.

</font>

---

In [20]:
#Q22

def extract_s_scores(list_xi):
    σ = np.std(list_xi)
    μ = np.mean(list_xi)
    XT = list_xi[len(list_xi)-1]
    S_score= (XT-μ)/σ
    return S_score

# Autoencoder Analysis

Autoencoders are neural networks used for unsupervised learning, particularly for dimensionality reduction and feature extraction. Training an autoencoder on the $Z_i$ matrix aims to identify hidden factors capturing the intrinsic structures in financial data.

### Architecture
- **Encoder**: Compresses input data into a smaller latent space representation.
  - *Input Layer*: Matches the number of features in the $Z_i$ matrix.
  - *Hidden Layers*: Compress data through progressively smaller layers.
  - *Latent Space*: Encodes the data into hidden factors.
- **Decoder**: Reconstructs input data from the latent space.
  - *Hidden Layers*: Gradually expand to the original dimension.
  - *Output Layer*: Matches the input layer to recreate the original matrix.

### Training
The autoencoder is trained by minimizing reconstruction loss, usually mean squared error (MSE), between the input $Z_i$ matrix and the decoder's output.

### Hidden Factors Extraction
After training, the encoder's latent space provides the most important underlying patterns in the stock returns.

---
<font color=green>Q23: (2 Marks) </font>
<br><font color='green'>
Modify the standardized returns matrix `Z_matrix` to reduce the influence of extreme outliers on model trainingby ensuring that all values in the matrix `Z_matrix` do not exceed 3 standard deviations from the mean. Specifically, cap these values at the interval $-3, 3]$. Store the adjusted values in a new matrix, `Z_hat`.
</font>

----

In [21]:
#Q23

Z_hat=Z_matrix

Z_hat[Z_hat > 3] = 3
Z_hat[Z_hat < -3] = -3

#Print the DataFrame as well as the minimum and maximum value in it to check if it is consistent with the question
print(Z_hat)
print("The minimum observed value of Z is: ", np.min(Z_hat))
print("The maximum observed value of Z is: ",np.max(Z_hat))

                ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2023-03-29  0.658925  2.189521  0.106285  0.207821  1.503570  0.444433   
2023-03-30  0.273051 -0.221306 -0.389329 -0.434766  0.787034  0.526996   
2023-03-31  0.360570  1.136308  1.540732  1.439604  0.531735 -0.056439   
2023-04-03 -0.713053 -2.259591  0.252735  0.407399 -0.591892 -0.600163   
2023-04-04  0.560730 -1.137430  0.099652  0.013877  0.658632 -0.342217   
...              ...       ...       ...       ...       ...       ...   
2024-03-22 -1.146803 -0.516681  1.151431  1.085070  0.074915  0.081847   
2024-03-25  0.659349 -1.221008 -0.372489 -0.341074  0.109668 -0.292706   
2024-03-26 -0.032708  0.234343  0.131651  0.109342 -0.556130 -0.244717   
2024-03-27 -0.363721  1.052056 -0.024166 -0.010591  0.315892  0.224881   
2024-03-28 -0.048375  0.411934 -0.078409  0.019962  0.022729  0.067776   

                 AEP      AMGN       

---
<font color=green>Q24: (1 Marks) </font>
<br><font color='green'>
Fetch the `Z_hat` data from GitHub, and we'll proceed with it now.
</font>



In [22]:
#Q24

Z_hat=pd.read_csv("https://raw.githubusercontent.com/Jandsy/ml_finance_imperial/main/Coursework/Z_hat.csv", index_col=0, parse_dates=True)

print(Z_hat)

                ADBE       ADP     GOOGL      GOOG      AMZN       AMD  \
Date                                                                     
2023-03-29  0.658925  2.189521  0.106285  0.207821  1.503570  0.444433   
2023-03-30  0.273051 -0.221306 -0.389329 -0.434766  0.787034  0.526996   
2023-03-31  0.360570  1.136308  1.540732  1.439604  0.531735 -0.056439   
2023-04-03 -0.713053 -2.259591  0.252735  0.407399 -0.591892 -0.600163   
2023-04-04  0.560730 -1.137430  0.099652  0.013877  0.658632 -0.342217   
...              ...       ...       ...       ...       ...       ...   
2024-03-22 -1.146803 -0.516681  1.151431  1.085070  0.074915  0.081847   
2024-03-25  0.659349 -1.221008 -0.372489 -0.341074  0.109668 -0.292706   
2024-03-26 -0.032708  0.234343  0.131651  0.109342 -0.556130 -0.244717   
2024-03-27 -0.363721  1.052056 -0.024166 -0.010591  0.315892  0.224881   
2024-03-28 -0.048375  0.411934 -0.078409  0.019962  0.022729  0.067776   

                 AEP      AMGN       

---
<font color=green>Q25: (3 Marks) </font>
<br><font color='green'>
Segment the standardized and capped returns matrix $\hat{Z}$ into two subsets for model training and testing. Precisly Allocate 70% of the data in $\hat{Z}$ to the training set $ \hat{Z}_{train} $ and Allocate the remaining 30% to the testing set $\hat{Z}_{test}$. Treat each stock within $\hat{Z}$ as an individual sample, by flattening temporal dependencies.
</font>



In [23]:
#Q25

import pandas as pd
from sklearn.model_selection import train_test_split

# Assuming Z_hat is your dataframe and the columns represent companies

# Calculate the threshold for 70% of the companies
threshold = int(0.7 * Z_hat.shape[1])

# Split into training and test data based on the columns (companies)
Z_hat_train = Z_hat.iloc[:, :threshold].reset_index(drop=False)
Z_hat_test = Z_hat.iloc[:, threshold:].reset_index(drop=False)

# Turn the 'Date' column to datetime and set it as the index
Z_hat_train["Date"] = pd.to_datetime(Z_hat_train["Date"])
Z_hat_train.set_index("Date", inplace=True)

Z_hat_test["Date"] = pd.to_datetime(Z_hat_test["Date"])
Z_hat_test.set_index("Date", inplace=True)

# Print the training and test sets as well as the shapes to double check
print('Z_hat_train dimensions', Z_hat_train.shape, 'and Z_hat_test dimensions' , Z_hat_test.shape)

Z_hat_train dimensions (252, 62) and Z_hat_test dimensions (252, 27)


---
<font color=green>Q26: (10 Marks) </font>
<br><font color='green'>
Please create an autoencoder following the instructions provided in  **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**, Use the model 'Variant 2' in Table 1.
</font>

---

In [24]:
#Q26

from tensorflow.keras import layers, models

#Determine the input shape from Z_hat by transposing

input_shape = Z_hat_train.T.shape[1] #Number of features in the input data

input_layer = layers.Input(shape=(input_shape,))
    
#Encoding layer (using tanh activation)
encoded = layers.Dense(20, activation='tanh', use_bias=True)(input_layer) 
    
#Decoding layer (using tanh activation)
decoded = layers.Dense(input_shape, activation='tanh', use_bias=True)(encoded)
    
#Define autoencoder models
autoencoder = models.Model(inputs=input_layer, outputs=decoded)

autoencoder_encoder = models.Model(input_layer, outputs=encoded)

---
<font color=green>Q27 (1 Mark) :

Display all the parameters of the deep neural network.
</font>

---

In [25]:
#Q27

# Create the autoencoder model
autoencoder.compile(optimizer='adam', loss='mean_squared_error')
print(autoencoder.summary())

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 252)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 20)             │         5,060 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 252)            │         5,292 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 10,352 (40.44 KB)

 Trainable params: 10,352 (40.44 KB)

 Non-trainable params: 0 (0.00 B)

None


---
<font color=green>Q28: (3 Marks) </font>
<br><font color='green'>
Train your model using the Adam optimizer for 20 epochs with a batch size equal to 8 and validation split to 20%. Specify the loss function you've chosen.
</font>


In [26]:
#Q28

# Train the autoencoder model for 20 epochs with batch size 8 and validation split of 20%
history = autoencoder.fit(Z_hat_train.T, Z_hat_train.T, epochs=20, batch_size=8, validation_split=0.2)

Epoch 1/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.9011 - val_loss: 0.8513
Epoch 2/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 0.8592 - val_loss: 0.8192
Epoch 3/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 0.8198 - val_loss: 0.7887
Epoch 4/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 0.7710 - val_loss: 0.7559
Epoch 5/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.7437 - val_loss: 0.7216
Epoch 6/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.7134 - val_loss: 0.6892
Epoch 7/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.6785 - val_loss: 0.6631
Epoch 8/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - loss: 0.6611 - val_loss: 0.6415
Epoch 9/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.6287 - val_loss: 0.6239
Epoch 10/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.6258 - val_loss: 0.6105
Epoch 11/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.6030 - val_loss: 0.6012
Epoch 12/20
7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.5871 - val_loss: 0.5937
Epoch 13/20
7

---
<font color=green>Q29: (3 Marks) </font>
<br><font color='green'>
Predict using the testing set and extract the residuals based on the methodology described in **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**.
for 'NVDA' stock.
</font>

---

In [27]:
# Find the index of the 'NVDA' column
nvda_index = Z_hat_test.columns.get_loc('NVDA')
print('NVDA column index:', nvda_index)
    
# Make the prediction by using the autoencoder on the test data and double-check the shapes
hidden_factors = autoencoder_encoder.predict(Z_hat_test.T)
print(hidden_factors.shape)
Z_pred_test = autoencoder.predict(Z_hat_test.T)
print(Z_pred_test.shape)

# Ensure residuals calculation is correct
residuals = Z_hat_test.T.values - Z_pred_test  

# Extract residuals for 'NVDA' stock using the correct index
nvda_residuals = residuals[nvda_index, :]

# Display the residuals for 'NVDA'
nvda_residuals_df = pd.DataFrame(nvda_residuals, columns=['NVDA Residuals'])
print(nvda_residuals_df)


NVDA column index: 0
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
(27, 20)
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step
(27, 252)
     NVDA Residuals
0         -0.015134
1         -0.202677
2         -0.056248
3          0.127158
4         -0.475291
..              ...
247        1.052305
248        0.584962
249       -1.081445
250       -1.437360
251        0.017784

[252 rows x 1 columns]


In [28]:
#Q29

# Find the index of the 'NVDA' column
nvda_index = Z_hat_test.columns.get_loc('NVDA')
print('NVDA column index:', nvda_index)
    
# Make the prediction by using the autoencoder on the test data and double-check the shapes
hidden_factors = autoencoder_encoder.predict(Z_hat_test.T)
print(hidden_factors.shape)
Z_pred_test = autoencoder.predict(Z_hat_test.T)
print(Z_pred_test.shape)

# Ensure residuals calculation is correct
residuals = Z_hat_test.T.values - Z_pred_test

# Extract residuals for 'NVDA' stock using the correct index
nvda_residuals = residuals[nvda_index, :]

# Display the residuals for 'NVDA'
nvda_residuals_df = pd.DataFrame(nvda_residuals, columns=['NVDA Residuals'])
print(nvda_residuals_df)


NVDA column index: 0
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step
(27, 20)
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step
(27, 252)
     NVDA Residuals
0         -0.015134
1         -0.202677
2         -0.056248
3          0.127158
4         -0.475291
..              ...
247        1.052305
248        0.584962
249       -1.081445
250       -1.437360
251        0.017784

[252 rows x 1 columns]


<font color=green>Q30: (7 Marks) </font>
<br><font color='green'>
By reading carrefully the paper **[End-to-End Policy Learning of a Statistical Arbitrage Autoencoder Architecture](https://arxiv.org/pdf/2402.08233.pdf)**, answers the following question:
1. **Summarize the Key Actions**: Highlight the main experiments and methodologies employed by the authors in Section 5.
2. **Reproduction Steps**: Detail the necessary steps required to replicate the authors' approach based on the descriptions provided in the paper.
3. **Proposed Improvement**: Suggest one potential enhancement to the methodology that could potentially increase the effectiveness or efficiency of the model.



**Summary of Key Actions:** 

The main idea of section 5 in the paper is to provide a detailed analysis of the steps followed that create the autoencoder strategy.

The authors propose the combination of asset pricing models along with statistical arbitrage parameters, that try to maximize the trading performance of the strategy.

The Autoencoder takes as an input the standardized stock returns $\hat{Z}$, uses a one layer encoder to reduce the dimensionality of the data, and then reconstructs the data into the initial dimension using a one layer decoder. 

$ \hat{Z}$  = $\tanh(W^{(1)}(relu(W^{(0)}X + b^{(0)})) + b^{(1)})$


Through the encoding phase the authors implement the reLu activation function, while during the decoding phase they apply the tanh activation function. Moreover, they propose the MSE as a Loss Function, using the Adam Optimizer for 10 Epochs. Once they calculate the reconstructed returns, the proceed by calculating the residuals.

$\epsilon_i(t) = Z_i(t) - \hat{Z}_i(t)$

Once, the residuals are calculated, they are then fed into a second neural network, that "constructs" portfolio weights, utilizing a hyperbolic tangent activation function, skiping a bias term.

$w_t = \tanh(W^{(2)}(\hat{Z}_t - Z_t)), \quad W^{(2)} \in \mathbb{R}^{N \times N}$

This second layer, introduces the constraint of Sharpe ratio, so the overall objective becomes to minimize the MSE, while on the same time maximize the Sharpe Ratio, with a parameter λ = 0.5

$\min_{\theta} \left( \lambda \text{MSE}(Z_t, \hat{Z}_t) + (1 - \lambda) \text{Sharpe}(w_t, R_{t+1}) \right)$


**Reproduction Steps:**

In order to replicate the approach that is described by the authors, we have to follow the following steps:

**1) Data Preperation:**
The Authors in the paper use daily returns for the US stock market. After the data has been processed, they proceed with calculating the Covariance Matrix over a lookback period of 252 trading days.

**2) Autoencoder Structure:**
The authors train an autoencoder, with the objective of minimizing the Mean Squared Error.

**3) Generating Residuals:**
From the reconstructed returns that the autoencoder provides, the authors calculate the residuals.

$\epsilon_i(t) = Z_i(t) - \hat{Z}_i(t)$

**4) Use the Ornstein-Uhlenbeck Process:**
Model the residuals, using the OU process, and estimate the model parameters such as $k_i$, $m_i$, and $\sigma_i$

$dX_i(t) = k_i (m_i - X_i(t)) \, dt + \sigma_i \, dW_i(t)$

After that, calculate the s-score in order to generate trading signals

$$
s_i(t) = \frac{X_i(t) - m_i}{\sigma_{eq,i}}
$$

where

$$
\sigma_{eq,i} = \frac{\sigma_i}{\sqrt{2k_i}}.
$$

**5) Generate Trading Signals:**

- **Open Long Position**: If $s_i(t) < -1.25$.
- **Open Short Position**: If $s_i(t) > 1.25$.
- **Close Long Position**: If $s_i(t) > -0.5$.
- **Close Short Position**: If $s_i(t) < 0.75$.

**6) Construct the portfolio:**
Determine the optimal weights by implementing:

$w_t = \tanh(W^{(2)}(\hat{Z}_t - Z_t))$

**7) Optimize the strategy:** Implement the Loss Function:  $\min_{\theta} \left( \lambda \text{MSE}(Z_t, \hat{Z}_t) + (1 - \lambda) \text{Sharpe}(w_t, R_{t+1}) \right)$

Train the model for the past 252 days, applying the Adam Optimizer for 10 epoch and a learning rate of 0.001.


**Proposed Improvement:**

One proposed enhancement by the authors introduces the modeling of transaction costs. Specifically, due to the nature of the model, there exists a high risk from the increased turnover and daily architectural changes.

This risk is showcased through the performance of sharpe ratio between the different strategies, since higher turnover increases the risk of the model. Although the Autoencoder delivers exceptionaly better returns compared to other models, if we take into consideration the sharpe ratio we can derive that the strategies perform equally adjusted to risk.

An example can be showcased through: AE OU 3 Variant 6  <
  PCA OU 8 Factors  ⇔
  0.42  <
  0.96
  
One proposed solution to mitigate the impact of transaction costs is to implement penalties for frequent rebalancing. By introducing such penalties, the strategy can be encouraged to make more significant and less frequent adjustments, thereby reducing turnover and associated costs. This approach can help balance the trade-off between optimizing returns and minimizing costs.